_drafts/arm-cortex-m-exceptions-and-nvic.md

```diff
@@ -0,0 +1,699 @@
+
+### Built in Exceptions
+
+These are exceptions that are part of every ARM Cortex-M core. The ARM Cortex-M specifications
+reserve **Exception Number**s **1**-**15**, inclusive, for these.
+
+> NOTE: Recall that the **Exception Number** maps to an offset within the **Vector Table**. Index 0 of
+> the Vector Table holds the reset value of the Main stack pointer. The rest of the **Vector Table**
+> holds **Exception Handler** pointers which is why the **Exception Number**s start at 1
+
+Six exception handlers are always supported and depending on the Cortex-M variant, additional
+handlers will be implemented as well. The minimum set is:
+
+- **Reset** - This is the routine executed when a chip comes out of reset. More details can be found within the zero-to-main bootloader series of posts[^2]
+- **Non Maskable Interrupt** (`NMI` ) - As the name implies, the interrupt cannot be disabled. If
+  errors happen in other exception handlers, a NMI will be triggered. Aside from the `Reset`
+  exception, it has the highest priority of all exceptions.
+- **HardFault** - The catchall for assorted system failures that can take place such as accesses to
+  bad memory, divide-by-zero errors and illegal unaligned accesses. It's the only handler for
+  faults on the ARMv6-M architecture but for ARMv7-M & ARMv8-M, finer granularity fault handlers
+  can be enabled for specific error classes (i.e `MemManage`, `BusFault`, `UsageFault`) [^3]
+- **SVCall** - Exception handler invoked when an _Supervisor Call_ (`svc`) instruction is executed
+- **PendSV** & **SysTick** - System level interrupts triggered by software. They are typically used
+  when running a RTOS to manage when the scheduler runs and when context switches take place.
+
+### External Interrupts
+
+ARM cores also support interrupt lines which are "external" to the core itself. These interrupt
+lines are usually routed to vendor-specific peripherals on the MCU such as Direct Memory Access (**DMA**)
+engines or General purpose input/output pins (**GPIO**s). All of these interrupts are configured via a peripheral known as the _Nested Vectored Interrupt Controller_ (**NVIC**).
+
+The **Exception Number** for external interrupts starts at **16**. The ARMv7-M reference manual has a good graphic which displays the Exception number mappings:
+
+![](img/armv-m-exceptions/exception-numbers.png)
+
+## Registers used to configure Cortex-M Exceptions
+
+Exceptions are configured on Cortex-M devices using a small set of registers within the **System
+Control Space** (**SCS**). An in-depth list of all the registers involved in exception handling can be found
+in the "Exception status and control" section of the ARMv7-M reference manual [^4]. A great way to
+build out an understanding of how the exception subsystem works is to walk through some of these. In the sections below we will explore the highlights.
+
+## Interrupt Control and State Register (ICSR) - 0xE000ED04
```

Comment:
haha yes, good catch, thats what I originally had ... we seem to have added a global kramdown setting that prevents H4s from showing up. I was trying to make it so all the registers (including the nested NVIC) ones could be seen from the TOC

```
# Markdown Settings
kramdown:
    toc_levels: 1..3
```

Googling around I don't see anyway to override this setting from an individual post. I realize it's a bit more tedious, but I think I'd favor just using `{:no_toc}` for sections that we want dropped for more control and dropping this default. Would need to audit a few old posts to figure out what needs to be disabled

Reply:
I changed that Kramdown setting to avoid super long TOCs. Feel free to revert it, the post to check is the one added with the commit where I changed the kramdown settings.

Reply:
sounds good, ill take a look, thanks!

Reply:
`###` here would work fine with current kramdown settings.

Reply:
It's the #### that did not work but I think I've changed my mind here, I'm going to slim down the TOC so its a little smaller.